File: compiler/feedback/attributes.go
package feedback

import (
	"wipple/database"
	"wipple/queries"
)

func registerAttributes() {
	register(Feedback[struct{}]{
		Id:    "extra-attribute-value",
		Rank:  RankSyntax,
		Query: queries.ExtraAttributeValue,
		Render: func(render *Render, node database.Node, data struct{}) {
			render.WriteNode(node)
			render.WriteString(" doesn't accept a value.")
			render.WriteBreak()
			render.WriteString("Try removing the value from this attribute.")
		},
	})

	register(Feedback[struct{}]{
		Id:    "duplicate-attribute",
		Rank:  RankSyntax,
		Query: queries.DuplicateAttribute,
		Render: func(render *Render, node database.Node, data struct{}) {
			render.WriteNode(node)
			render.WriteString(" is defined more than once.")
			render.WriteBreak()
			render.WriteString("Try removing this attribute.")
		},
	})

	register(Feedback[struct{}]{
		Id:    "missing-attribute-value",
		Rank:  RankSyntax,
		Query: queries.MissingAttributeValue,
		Render: func(render *Render, node database.Node, data struct{}) {
			render.WriteNode(node)
			render.WriteString(" is missing a value.")
			render.WriteBreak()
			render.WriteString("Try adding a value to this attribute using ")
			render.WriteCode(":")
			render.WriteString(".")
		},
	})
}

File: compiler/src/feedback/items/attributes.rs
use crate::{
    feedback::{FeedbackCtx, FeedbackRank, RegisteredFeedback},
    nodes::{DuplicateAttributeValue, ExtraAttributeValue, MissingAttributeValue},
    queries,
};
use std::collections::BTreeSet;

pub fn register(ctx: &mut FeedbackCtx) {
    ctx.register(RegisteredFeedback::new(
        "extra-attribute-value",
        queries::fact::<ExtraAttributeValue>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, _)| {
            w.write_node(node);
            w.write_string(" doesn't accept a value.");
            w.write_break();
            w.write_string("Try removing the value from this attribute.");
        },
    ));

    ctx.register(RegisteredFeedback::new(
        "duplicate-attribute",
        queries::fact::<DuplicateAttributeValue>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, _)| {
            w.write_node(node);
            w.write_string(" is defined more than once.");
            w.write_break();
            w.write_string("Try removing this attribute.");
        },
    ));

    ctx.register(RegisteredFeedback::new(
        "missing-attribute-value",
        queries::fact::<MissingAttributeValue>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, _)| {
            w.write_node(node);
            w.write_string(" is missing a value.");
            w.write_break();
            w.write_string("Try adding a value to this attribute using ");
            w.write_code(":");
            w.write_string(".");
        },
    ));
}
